src/ripple/shamap/README.md

```diff
@@ -26,164 +56,292 @@ There are two different ways of building and using a SHAMap:
+SHAMap, it first clones it and then sets the new copy to have a sequence number
+equal to the SHAMap sequence number.  The `unshareNode` is a private utility
+which automates the task of first checking if the node is already sharable, and
+if so, cloning it and giving it the proper sequence number.  An example case
+where a private copy is needed is when an inner node needs to have a child
+pointer altered.  Any modification to a node will require a non-shared node.
+
+When a SHAMap decides that it is safe to share a node of its own, it sets the
+node's sequence number to 0 (a SHAMap never has a sequence number of 0). This
+is done for every node in the trie when `SHAMap::walkSubTree` is executed.
+
+Note that other objects in rippled also have sequence numbers (e.g. ledgers).
+The SHAMap and node sequence numbers should not be confused with these other
+sequence numbers (no relation).
 
 ## SHAMap Creation ##
 
 A SHAMap is usually not created from vacuum.  Once an initial SHAMap is
-constructed, later SHAMaps are usually created by calling
-snapShot(bool isMutable) on the original SHAMap().  The returned SHAMap
-has the expected characteristics (mutable or immutable) based on the passed
-in flag.
+constructed, later SHAMaps are usually created by calling snapShot(bool
+isMutable) on the original SHAMap().  The returned SHAMap has the expected
+characteristics (mutable or immutable) based on the passed in flag.
 
 It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
-snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might
-be modified must be copied before they are placed in the mutable map.
+snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might be
+modified must be copied before they are placed in the mutable map.
+
+A new SHAMap is created with each new ledger round.  Transactions not executed
+in the previous ledger populate the SHAMap for the new ledger.
+
+## Storing SHAMap data in the database ##
+
+When consensus is reached, the ledger is closed.  As part of this process, the
+SHAMap is stored to the database by calling `SHAMap::flushDirty`.
+
+Both `unshare()` and `flushDirty` walk the SHAMap by calling
+`SHAMap::walkSubTree`.  As `unshare()` walks the trie, nodes are not written to
+the database, and as `flushDirty` walks the trie nodes are written to the
+database. `walkSubTree` visits every node in the trie. This process must ensure
+that each node is only owned by this trie, and so "unshares" as it walks each
+node (from the root down).  This is done in the `preFlushNode` function by
+setting the ensuring that the node has a sequence number equal to that of the
+SHAMap.  If the node doesn't, it is cloned.
+
+For each inner node encountered (starting with the root node), each of the
+children are inspected (from 1 to 16).  For each child, if it has a non-zero
+sequence number (unshareable), the child is first COW'd.  Then if the child is
```

Comment:
I know that COW is copy-on-write (though I prefer it spelled out at least once in formal documentation), but how is "COWed" different from "copied"? Is there a wrapper type to handle the COW? Does this paragraph imply that there is no concurrent access to the same `SHAMap`?